### PR TITLE
cql3: support list subscripts in WHERE clause

### PR DESCRIPTION
**Pull request by avikivity**
I noticed that `column_condition` (used in LWT `IF` clause) supports lists.
As part of the Grand Expression Unification we'll need to migrate that to
expressions, so we'll need to support list subscripts.

Use the opportunity to relax the normal filtering to allow filtering on
list subscripts: `WHERE my_list[:index] = :value`.

### CODE REVIEW

File: cql3/expr/expression.cc
Comment (psarna) [inline cql3/expr/expression.cc:243]: Something's extremely off with braces here and indentation later... I think it only works because the `else` body is a one-liner and you skipped a brace+indentation below.

**Review by psarna — COMMENTED** [review]
Looks good, left one remark

**Comment by benipeled** [comment]
> @benipeled why did continuous integration fail?

@avikivity it's the docs CI (run by github-actions) complaining about
```
writing output... [100%] index
/home/runner/.cache/pypoetry/virtualenvs/scylla-0iFb59df-py3.7/lib/python3.7/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document

  warn("Container node skipped: type={0}".format(mdnode.t))
/home/runner/work/scylla/scylla/docs/guides/debugging.md:152: WARNING: None:any reference target not found: ../../gdbinit
generating indices... genindex done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 1 warning.
sitemap.xml was generated for URL https://scylla.docs.scylladb.com/ in /home/runner/work/scylla/scylla/docs/_build/dirhtml/sitemap.xml
make: *** [Makefile:88: test] Error 1
make: Leaving directory '/home/runner/work/scylla/scylla/docs'
Error: Process completed with exit code 2.
```
I guess it has nothing with your change but some previous changes merged regardless of the CI results.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc:243]: Ouch! will fix.

**Comment by avikivity** [comment]
v2:
  - fix bad bracing in expression.cc get_value() (did not affect anything)
  - add reindent patch
  - add tests

**Comment by avikivity** [comment]
> > @benipeled why did continuous integration fail?
> 
> @avikivity it's the docs CI (run by github-actions) complaining about
> 
> ```
> writing output... [100%] index
> /home/runner/.cache/pypoetry/virtualenvs/scylla-0iFb59df-py3.7/lib/python3.7/site-packages/recommonmark/parser.py:75: UserWarning: Container node skipped: type=document
> 
>   warn("Container node skipped: type={0}".format(mdnode.t))
> /home/runner/work/scylla/scylla/docs/guides/debugging.md:152: WARNING: None:any reference target not found: ../../gdbinit
> generating indices... genindex done
> writing additional pages... search done
> copying static files... done
> copying extra files... done
> dumping search index in English (code: en)... done
> dumping object inventory... done
> build finished with problems, 1 warning.
> sitemap.xml was generated for URL https://scylla.docs.scylladb.com/ in /home/runner/work/scylla/scylla/docs/_build/dirhtml/sitemap.xml
> make: *** [Makefile:88: test] Error 1
> make: Leaving directory '/home/runner/work/scylla/scylla/docs'
> Error: Process completed with exit code 2.
> ```
> 
> I guess it has nothing with your change but some previous changes merged regardless of the CI results.

Can we fix it? These false positives make it hard to trust CI.

File: cql3/expr/expression.cc
Comment (avikivity) [reply cql3/expr/expression.cc]: Fixed. I remember feeling something was odd.

**Comment by benipeled** [comment]
> Can we fix it? These false positives make it hard to trust CI.

I'm not familiar with sphinx rules but it seems like the reference to a file with no extension causes this "error",
but AFAICS the `gdbinit` is an example file with mixed content,
@bhalevy can I move the `gdbinit` content to a markdown-block directly on the `debugging.md` file?

**Comment by avikivity** [comment]
`gdbinit` is not under docs, so why is it considered?

**Comment by avikivity** [comment]
v3: corrected the last test (problem in test)

**Comment by benipeled** [comment]
> @avikivity the gdbinit issue is described by me in this (also unrelated) PR: [#10426 (comment)](https://github.com/scylladb/scylla/pull/10426#issuecomment-1131420159) . It's sad, because the relative path in the docs is correct, it just happens not to be resolvable in the test environment - so it's just tests that are faulty.

As far as I can see, it causes a broken link on the docs page - https://scylla.docs.scylladb.com/master/guides/debugging.html (look for "Scylla provides the following gdbinit")

**Comment by avikivity** [comment]
#10650

File: cql3/expr/prepare_expr.cc
Comment (cvybhu) [inline cql3/expr/prepare_expr.cc:1004]: The error message is changed before the test that checks it. Because of it tests fail and it's not bisectable.

File: cql3/expr/to_restriction.cc
Comment (cvybhu) [inline cql3/expr/to_restriction.cc:72]: It would be good to change other error messages to include lists alongside maps.
Right now if someone attempts to make an indexed restriction on non-frozen list they will get a message about maps.

**Review by cvybhu — CHANGES_REQUESTED** [review]

File: cql3/expr/prepare_expr.cc
Comment (avikivity) [reply cql3/expr/prepare_expr.cc:1004]: Right, fixed.

File: cql3/expr/to_restriction.cc
Comment (avikivity) [reply cql3/expr/to_restriction.cc:72]: Adjusted. I think those restrictions are wrong btw. When filtering we can easily support non-equality relations or frozen collections. This check belongs in query planning when trying to select the index, not in validation.

**Comment by avikivity** [comment]
v4:
 - change test verifying 'cannot be used as a map' error message to new error message
 - change dubious validation messages that the collection is not frozen and that the relation is equality to reflect that lists are supported
 - rebased
 - tested bisectability

**Review by psarna — APPROVED** [review]

**Comment by psarna** [comment]
sigh, re-restarting...

File: cql3/expr/to_restriction.cc
Comment (nyh) [reply cql3/expr/to_restriction.cc:72]: E.g., https://github.com/scylladb/scylla/issues/9777

**Comment by avikivity** [comment]
@benipeled what can we do to fix this? CI is not continuous if we have to continually restart it.

**Comment by psarna** [comment]
as a quick workaround, we could just rename it for the time being

File: test/cql-pytest/test_filtering.py
Comment (nyh) [inline test/cql-pytest/test_filtering.py:258]: Good you tested that. Would also be nice to check that a null parameter, l[1] = null, does not cause a mess.
Also, please add checks for error conditions - e.g., `l[3] = 3` should complain that 3 is the wrong type, and so should `l['hello'] = 'hi'`.

File: test/cql-pytest/test_filtering.py
Comment (nyh) [inline test/cql-pytest/test_filtering.py:189]: There's some value in making the list check a *separate* test, because it is specifically a feature not supported by Cassandra and deserves its own comment, and "scylla_only" tag - it's not a Cassandra bug per se, but a feature that they simply don't have.
However, I don't feel very strongly about this. What you did is also good enough.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good, I just left a minor request for testing the error cases.
By the way you can update the commit message not to say that tests are missing.

**Comment by avikivity** [comment]
v3:
 - added three tests:
   - bad list index type
   - bad right-hand-side of binary operator involving subscripted list
   - NULL list index
- adjusted cover letter to remove completed TODOs

**Comment by benipeled** [comment]
> @benipeled what can we do to fix this? CI is not continuous if we have to continually restart it.

This is not a spot-termination but a DNS issue https://github.com/scylladb/scylla-pkg/issues/2888
```
docker: error pulling image configuration: Get
.
.
.
dial tcp: lookup docker-images-prod.s3.dualstack.us-east-1.amazonaws.com: no such host.
```

**Comment by benipeled** [comment]
@avikivity On Sunday I added three new Jenkins nodes - spider[1-3] - they're pretty busy with CI jobs and so far it looks great.

https://github.com/scylladb/scylla-pkg/pull/2909

**Review by nyh — APPROVED** [review]
Looks good. Waiting for the last CI run before merging.

**Pull request closed without merging**
